### PR TITLE
Add BaseAgent class

### PR DESCRIPTION
**Pull request by Het-Shah**
Added BaseAgent Class as of now. Will add code that will similar to all agents. I just created the basic structure for the class as of now.

### CODE REVIEW

**Comment by Het-Shah** [comment]
@Sharad24 for `collect_rollouts()` method in VPG, PPO, etc the 2048 number should be replaced by `rollout_size` right?

**Comment by Sharad24** [comment]
Yes, I missed that in my PR. Thanks!

**Comment by Het-Shah** [comment]
Cool I'll make the change in this PR.

File: genrl/deep/agents/vpg/vpg.py [hunk 1]
```diff
@@ -181,6 +181,7 @@ def get_traj_loss(self, value, done) -> None:
 
     def update_policy(self) -> None:
 
+        # what  is 256 
         for rollout in self.rollout.get(256):
```
Comment (Het-Shah) [inline genrl/deep/agents/vpg/vpg.py:185]: what is this 256? Is it the update interval?

**Comment by Sharad24** [comment]
Should be changed to batch size.

**Comment by Het-Shah** [comment]
Is it actor_batch_size?

**Comment by Het-Shah** [comment]
@Sharad24 Can you check the ppo and vpg implementation once. I can't think of any other stuff that we can abstract.

File: genrl/deep/common/base_class.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        raise NotImplementedError()
+
+    def get_hyperparameters(self) -> Dict[str, Any]:
+        """
+        Hyperparameters you want to return
+        """
+        raise NotImplementedError()
+
+
+class BaseOnPolicyAgent(BaseAgent):
```
Comment (Sharad24) [inline genrl/deep/common/base_class.py:93]: I think this can be renamed to just `OnPolicyAgent`?

File: genrl/deep/common/base_class.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                values.detach(),
+                old_log_probs,
+            )
+
+            state = next_state
+
+            for i, done in enumerate(dones):
+                if done:
+                    self.rewards.append(self.epoch_reward[i])
+                    self.epoch_reward[i] = 0
```
Comment (Sharad24) [inline genrl/deep/common/base_class.py:162]: `nit`

I think we can move this to another function to reduce complexity.

File: genrl/deep/common/base_class.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        self.load_model = load_model
+        self.save_interval = save_interval
+        self.observation_space = None
+        self.action_space = None
+
+        # Assign device
+        if "cuda" in device and torch.cuda.is_available():
+            self.device = torch.device(device)
+        else:
+            self.device = torch.device("cpu")
```
Comment (Sharad24) [inline genrl/deep/common/base_class.py:40]: I think we can define a method `to(self, device)` to transfer anything to the the specified device?

Just like pytorch I guess.

File: genrl/deep/common/base_class.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        raise NotImplementedError()
+
+    def _update_params_before_select_action(self, timestep: int) -> None:
+        """
+        Update any parameters before selecting action like epsilon for decaying epsilon greedy
+
+        :param timestep: Timestep in the training process
+        :type timestep: int
+        """
+        pass
```
Comment (Sharad24) [inline genrl/deep/common/base_class.py:60]: Should raise a notimplementederror?

File: genrl/deep/common/base_class.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        """
+        raise NotImplementedError()
+
+
+class BaseOnPolicyAgent(BaseAgent):
+    def __init__(
+        self,
+        network_type: str,
+        env: Any,
+        timesteps_per_actorbatch: int = 256,
```
Comment (Sharad24) [inline genrl/deep/common/base_class.py:98]: I think we can just rename this to `batch_size` or something.

File: genrl/deep/common/base_class.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        actor_batch_size: int = 64,
+        epochs: int = 100,
+        seed: Optional[int] = None,
+        render: bool = False,
+        device: Union[torch.device, str] = "cpu",
+        run_num: int = None,
+        save_model: str = None,
+        load_model: str = None,
+        save_interval: int = 50,
+        rollout_size: int = 2048,
```
Comment (Sharad24) [inline genrl/deep/common/base_class.py:112]: Some can be replaced with `*args, **kwargs`

File: genrl/deep/agents/vpg/vpg_temp.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+            seed,
+            render,
+            device,
+            run_num,
+            save_model,
+            load_model,
+            save_interval,
+            rollout_size,
+        )
+        self.policy_copy_interval = policy_copy_interval
```
Comment (Sharad24) [inline genrl/deep/agents/vpg/vpg_temp.py:63]: Not required in VPG?

File: genrl/deep/agents/vpg/vpg_temp.py [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+
+        self.create_model()
```
Comment (Sharad24) [inline genrl/deep/agents/vpg/vpg_temp.py:65]: We should a bool argument like `setup_model` per basis we choose to call this method, otherwise its calling this every time

File: genrl/deep/common/base_class.py (hunk 2 above)
Comment (Het-Shah) [reply genrl/deep/common/base_class.py:93]: Ok

File: genrl/deep/common/base_class.py (hunk 4 above)
Comment (Het-Shah) [reply genrl/deep/common/base_class.py:40]: Ok

File: genrl/deep/common/base_class.py (hunk 5 above)
Comment (Het-Shah) [reply genrl/deep/common/base_class.py:60]: No, because it won't be defined in every agent.

File: genrl/deep/common/base_class.py (hunk 6 above)
Comment (Het-Shah) [reply genrl/deep/common/base_class.py:98]: I'll rename it to `batch_size`

File: genrl/deep/agents/vpg/vpg_temp.py (hunk 8 above)
Comment (Het-Shah) [reply genrl/deep/agents/vpg/vpg_temp.py:63]: I just copied every parameter from the original VPG, will check if its used

File: genrl/deep/agents/ppo1/ppo_temp.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+
+class PPO1(BaseOnPolicyAgent):
+    def __init__(
+        self,
+        network_type: str,
+        env: Union[gym.Env, VecEnv],
+        timesteps_per_actorbatch: int = 256,
+        gamma: float = 0.99,
+        clip_param: float = 0.2,
+        actor_batch_size: int = 64,
```
Comment (Het-Shah) [inline genrl/deep/agents/ppo1/ppo_temp.py:30]: Do we even need this parameter? I don't see it being used anywhere.

File: genrl/deep/common/base_class.py (hunk 7 above)
Comment (Het-Shah) [reply genrl/deep/common/base_class.py:112]: I don't really think it'll be the correct thing here, so not doing this as of now, but if I think it might help in future then I'll add it.

Comment (Sharad24) [reply genrl/deep/common/base_class.py:112]: Why wouldn't it be correct?

File: genrl/deep/agents/vpg/vpg_temp.py (hunk 9 above)
Comment (Sharad24) [reply genrl/deep/agents/vpg/vpg_temp.py:56]: Yeah, so I was saying for every time the agent is initialized, it is called. Maybe we can do it in another issue, nvm.

File: genrl/deep/agents/ppo1/ppo_temp.py [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+                rollout.observations, actions
+            )
+
+            advantages = rollout.advantages
+            advantages = (advantages - advantages.mean()) / (advantages.std() + 1e-8)
+
+            ratio = torch.exp(log_prob - rollout.old_log_prob)
+
+            policy_loss_1 = advantages * ratio
+            policy_loss_2 = advantages * torch.clamp(ratio, 1 - 0.2, 1 + 0.2)
```
Comment (Sharad24) [inline genrl/deep/agents/ppo1/ppo_temp.py:135]: Clip param should be used here, instead of 0.2

File: genrl/deep/agents/ppo1/ppo_temp.py (hunk 10 above)
Comment (Sharad24) [reply genrl/deep/agents/ppo1/ppo_temp.py:21]: Actor_batch_size?

File: genrl/deep/common/base_class.py (hunk 7 above)
Comment (Het-Shah) [reply genrl/deep/common/base_class.py:112]: Yea coz we need these all params to be there, we generally use `**kwargs` for optional parameters no? (Only seed, device seem optional to me)

File: genrl/deep/agents/ppo1/ppo_temp.py (hunk 11 above)
Comment (Het-Shah) [reply genrl/deep/agents/ppo1/ppo_temp.py:135]: Ok will make this change.

File: genrl/deep/agents/ppo1/ppo_temp.py (hunk 10 above)
Comment (Het-Shah) [reply genrl/deep/agents/ppo1/ppo_temp.py:21]: Yes

File: genrl/deep/common/base_class.py (hunk 7 above)
Comment (Sharad24) [reply genrl/deep/common/base_class.py:112]: No, I don't think there is anything about it being for optional parameters

File: genrl/deep/agents/ppo1/ppo_temp.py (hunk 10 above)
Comment (Sharad24) [reply genrl/deep/agents/ppo1/ppo_temp.py:21]: Yeah probably not after batch size

File: genrl/deep/common/base_class.py (hunk 7 above)
Comment (Het-Shah) [reply genrl/deep/common/base_class.py:112]: Yea but if we have parameters that are used majorly we don't generally put it in `**kwargs`, this is what I believe :|

File: genrl/deep/agents/ppo1/ppo_temp.py (hunk 10 above)
Comment (Het-Shah) [reply genrl/deep/agents/ppo1/ppo_temp.py:21]: It wasn't use anywhere even before, I'll remove this.

**Comment by Het-Shah** [comment]
OnPolicy is done majorly. It passes tests as well. Will move to OffPolicy.

**Comment by sampreet-arthi** [comment]
Also, were you thinking about doing #152 ? If not, I guess the issue can be closed ig.

File: genrl/deep/common/base_class.py (hunk 7 above)
Comment (Sharad24) [reply genrl/deep/common/base_class.py:112]: Why not? **kwargs will just store the parameters it cannot find in the existing keyword arguments (i.e., the ones inherited) and pass to superclasses.

Another point is it would reduce code climate issues

@ajaysub110 What are your thoughts on this?

Comment (ajaysub110) [reply genrl/deep/common/base_class.py:112]: While **kwargs will clean stuff up, I think it will affect our docs. Its kind of annoying when the docs says args, kwargs and you're required to see the code examples to identify possible arguments to the class, no?

Comment (Sharad24) [reply genrl/deep/common/base_class.py:112]: Right so should keep the doctoring intact but remove arguments which can be, with args and kwargs?

Comment (ajaysub110) [reply genrl/deep/common/base_class.py:112]: Yeah, that makes sense

**Comment by Het-Shah** [comment]
Added `**kwargs`. Check if it is fine @Sharad24 @ajaysub110

File: genrl/deep/agents/a2c/a2c_temp.py [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+
+from ....environments.vec_env import VecEnv
+from ...common import OnPolicyAgent, RolloutBuffer, get_env_properties, get_model
+
+
+class A2C(OnPolicyAgent):
+    def __init__(
+        self,
+        network_type: str,
+        env: Union[gym.Env, VecEnv],
```
Comment (Sharad24) [inline genrl/deep/agents/a2c/a2c_temp.py:18]: This is same in all agents, so can be extracted out with *args I guess

Comment (Het-Shah) [reply genrl/deep/agents/a2c/a2c_temp.py:18]: Yea I was thinking of this initially but then these are really important variables, I think should be kept out.

File: genrl/deep/common/base_class.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+    def __init__(self, network_type: str, env: Any, epochs: int = 100, **kwargs):
+        self.network_type = network_type
+        self.env = env
+        self.epochs = epochs
+        self.seed = kwargs.get("seed", None)
+        self.render = kwargs.get("render", False)
+        self.run_num = kwargs.get("run_num", None)
+        self.save_model = kwargs.get("save_model", None)
+        self.load_model = kwargs.get("load_model", None)
+        self.save_interval = kwargs.get("save_interval", 50)
```
Comment (Sharad24) [inline genrl/deep/common/base_class.py:20]: Are you setting this way for a particular reason?

Comment (Het-Shah) [reply genrl/deep/common/base_class.py:20]: Yes this sets a default value to the variables. So if we are using it somewhere in the code unknowingly to it doesn't throw errors.

Comment (Sharad24) [reply genrl/deep/common/base_class.py:20]: Wouldn't that happen automatically if we specify the default value in the function definition?

Comment (Het-Shah) [reply genrl/deep/common/base_class.py:20]: yes, but this was to  be on the safer side

Comment (Het-Shah) [reply genrl/deep/common/base_class.py:20]: won't really make a big difference tbh

Comment (Sharad24) [reply genrl/deep/common/base_class.py:20]: Right, just saying because its not a standard way to do it

**Comment by Het-Shah** [comment]
Codeclimate Issues are more or less fixed now.

**Comment by Sharad24** [comment]
@ajaysub110 Do you have time to review once?

**Comment by Het-Shah** [comment]
It is showing build canceled, because of the docs. :(

**Comment by Sharad24** [comment]
Restarted the build

**Comment by Sharad24** [comment]
Can you rename `base_class.py` to `base.py` and move it to `agents/`, probably makes more sense there.

**Comment by Sharad24** [comment]
Have you tried training the algos once? For onpolicy trainer, these worked for instance - 
1. PPO w/ CartPole, Lunarlander, Acrobot
2. VPG w/ CartPole
3. A2C w/ CartPole, Acrobot

**Comment by ajaysub110** [comment]
I think I can, but only a little later if that's okay.

**Comment by Sharad24** [comment]
No problem, nature boi

**Comment by Het-Shah** [comment]
> Can you rename base_class.py to base.py and move it to agents/, probably makes more sense there.

Yea I'll do this. 

>Have you tried training the algos once? For onpolicy trainer, these worked for instance -

I checked CartPole for every agent for 10 epochs. I can't really train for longer epochs though my ubuntu is a bit fucked, it hangs and I have to reboot. Can you check and let me know if it doesn't train. :(

**Comment by Sharad24** [comment]
@ajaysub110 Are you going to review?

**Comment by ajaysub110** [comment]
Already had a look yesterday. Sorry, I had approved on the group

**Comment by Het-Shah** [comment]
I'll resolve the merge conflicts

File: genrl/deep/agents/vpg/vpg.py [hunk 14]
```diff
@@ -56,6 +56,7 @@ def __init__(
         gamma: float = 0.99,
         epochs: int = 1000,
         lr_policy: float = 0.01,
+        lr_value: float = 0.5,
```
Comment (Sharad24) [inline genrl/deep/agents/vpg/vpg.py:59]: Lr value should not be needed in VPG, I think

Comment (Het-Shah) [reply genrl/deep/agents/vpg/vpg.py:59]: I'll need to change it.

**Comment by Het-Shah** [comment]
I did the logger thing here only wasn't a big change.

**Comment by Het-Shah** [comment]
Ready to merge.

File: genrl/deep/agents/a2c/a2c.py [hunk 15]
```diff
@@ -313,7 +214,6 @@ def get_hyperparams(self) -> Dict[str, Any]:
             "network_type": self.network_type,
             "timesteps_per_actorbatch": self.timesteps_per_actorbatch,
             "gamma": self.gamma,
-            "actor_batch_size": self.actor_batch_size,
             "lr_actor": self.lr_actor,
             "lr_critic": self.lr_critic,
             "actor_weights": self.ac.actor.state_dict(),
```
Comment (sampreet-arthi) [inline genrl/deep/agents/a2c/a2c.py:219]: Shouldn't we save `batch_size` and `rollout_size` too?

File: genrl/deep/agents/base.py [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+
+    def collect_rewards(self, dones):
+        for i, done in enumerate(dones):
+            if done:
+                self.rewards.append(self.epoch_reward[i])
+                self.epoch_reward[i] = 0
+
+    def collect_rollouts(self, initial_state):
+
+        state = initial_state
```
Comment (sampreet-arthi) [inline genrl/deep/agents/base.py:114]: Simpler to just use `def collect_rollouts(self, state)` ?

File: genrl/deep/agents/ddpg/ddpg.py [hunk 17]
```diff
@@ -293,31 +289,31 @@ def update_params(self, update_interval: int) -> None:
...
-            self.optimizer_q.step()
-            self.optimizer_policy.step()
+            # loss = self.get_q_loss(state, action, reward, next_state, done)
+            # loss.backward()
 
-            # self.optimizer_q.zero_grad()
-            # loss_q = self.get_q_loss(state, action, reward, next_state, done)
-            # loss_q.backward()
             # self.optimizer_q.step()
+            # self.optimizer_policy.step()
```
Comment (sampreet-arthi) [inline genrl/deep/agents/ddpg/ddpg.py:302]: There are too many comments. Makes it more confusing as they're actually used later on.

File: genrl/deep/agents/ppo1/ppo1.py [hunk 18]
```diff
@@ -221,71 +187,12 @@ def update_policy(self):
...
         hyperparams = {
             "network_type": self.network_type,
-            "timesteps_per_actorbatch": self.timesteps_per_actorbatch,
+            "batch_size": self.batch_size,
             "gamma": self.gamma,
             "clip_param": self.clip_param,
-            "actor_batch_size": self.actor_batch_size,
             "lr_policy": self.lr_policy,
             "lr_value": self.lr_value,
             "policy_weights": self.policy_new.state_dict(),
```
Comment (sampreet-arthi) [inline genrl/deep/agents/ppo1/ppo1.py:198]: Save `rollout_size` here too. Atari envs will need a different rollout size so needs to be saved separately.

File: genrl/deep/agents/ddpg/ddpg.py (hunk 17 above)
Comment (Het-Shah) [reply genrl/deep/agents/ddpg/ddpg.py:302]: I'll remove the comments.

File: genrl/deep/agents/ppo1/ppo1.py (hunk 18 above)
Comment (Het-Shah) [reply genrl/deep/agents/ppo1/ppo1.py:198]: cool

**Comment by Het-Shah** [comment]
> As for the max_ep_len, I think it would be better if I add it along with my PR as TimeLimit has only been implemented in that.

Yea cool. I am not making this change here then.

**Comment by Sharad24** [comment]
Awesome, merging.
